docs: WIP-101 - RP Request Authorization Method for Smart Contracts (#594)
worldcoin/world-id-protocol: docs/WIPs/wip-101.md

---

## Abstract

This spec introduces a method to verify Relying Party (RP) proof requests in the World ID Protocol through on-chain smart contracts. As some RPs may not have a backend and be simply a contract, this allows OPRF Nodes to answer **"Does the RP authorize this request?"** in order to process relevant OPRF requests such as for nullifier generation.

## Motivation

**Context**: Proof requests from Relying Parties in the World ID Protocol MUST be signed in order to be recognized. This serves to ensure proofs are authorized for the intended recipient and reduce the attack surface in post-compromise scenarios for Authenticators.

A Relying Party MAY be a Smart Contract without a backend. When this is the case, having a private key to sign the request is impractical or even impossible in a secure way. Hence, we need a way for Relying Parties to authorize the request in a way compatible with Smart Contracts.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).

> The term "contract" is used to refer to "Smart Contracts", both may be used interchangeably.

Reviewer: nit: I don't think we capitalize smart contracts?
```suggestion
In this document, "contract" and "smart contract" are used interchangeably.
```

```mermaid
sequenceDiagram
participant c as RP Client (e.g. Web App)
participant a as Authenticator
participant o as OPRF Nodes
participant v as RP WIP101 Verifier

c->>c: generate rnd nonce
c->>c: encode RP custom `data`
c->>c: construct proof request
c->>a: proof request (unsigned)
a->>o: nullifier request
o->>v: verifyRpRequest()
v->>v: internal verifications
v->>o: is valid
o->>a: nullifier
a->>c: proof response
c->>v: submit proof and perform action on-chain
v->>v: invalidate nonce
```

## Recommended Validations

It is RECOMMENDED to implement the following validations for a Proof request:

> 1. Rreject a request which `version` is larger than the maximum known, as future version changes MAY be incompatible with previous versions.

Reviewer: ```suggestion
1. Reject a request which `version` is larger than the maximum known, as future version changes MAY be incompatible with previous versions.
```
